### PR TITLE
Add compression gzip

### PR DESCRIPTION
**Pull request by zmstone**

### CODE REVIEW

**Comment by zmstone** [comment]
@dgriff1 i have pulled your commits into this pr.
please send another pr targeting this branch if you have snappy ready.

File: src/kpro.erl [hunk 1]
```diff
@@ -333,14 +360,21 @@ decode(StructName, Bin) when is_atom(StructName) ->
   kpro_structs:decode(StructName, Bin).
 
 decode_message_stream(<<>>, Acc) ->
-  lists:reverse(Acc);
+  Acc;
```
Comment (id) [inline src/kpro.erl:372]: why this change?

Comment (zmstone) [inline src/kpro.erl:372]: because without this change the compressed messages would be reversed twice.
see line 373

Comment (id) [inline src/kpro.erl:372]: Right, you do the final reverse on 349.

**Comment by zmstone** [comment]
done.

**Comment by zmstone** [comment]
not going support lz4 in this pr.

File: src/kpro.erl [hunk 2]
```diff
@@ -333,14 +370,23 @@ decode(StructName, Bin) when is_atom(StructName) ->
...
     catch error : {badmatch, _} ->
       {?incomplete_message, <<>>}
     end,
-  decode_message_stream(Rest, [Msg | Acc]).
+  NewAcc =
+    case Msg of
+      #kpro_Message{attributes = Attr} = Msg when ?KPRO_IS_GZIP_ATTR(Attr) ->
+        decode_message_stream(zlib:gunzip(Msg#kpro_Message.value), Acc);
+      #kpro_Message{attributes = Attr} = Msg when ?KPRO_IS_SNAPPY_ATTR(Attr) ->
+        decode_message_stream(java_snappy_unpack(Msg#kpro_Message.value), Acc);
```
Comment (id) [inline src/kpro.erl:384]: Will it work with messages produced with brod/snappy?

Comment (id) [inline src/kpro.erl:384]: All right, so kafka always decompresses and compresses messages on its own and we always get java-snappy encoding in consumers?

Comment (zmstone) [inline src/kpro.erl:384]: yes, kafka can handle raw compressed messages (without chunking).
and you are right, it decompresses messages, and compress back with offsets assigned, then write to disk.

**Comment by id** [comment]
Good job 💯

**Comment by dgriff1** [comment]
@zmstone thank for finishing this. I didnt realize kafka wasnt using a standard snappy which explains why my snappy consumer always said the data was corrupt. I did find a kafka ticket yesterday saying they were going to add in standard lz4 which would make adding lz4 support far easier.

**Comment by zmstone** [comment]
@dgriff1 we are working on a native Erlang snappy lib, but that might take a bit longer.
will merge this pr without snappy and catch up later.

**Comment by dgriff1** [comment]
@zmstone was the deciding factor avoid NIFs?

**Comment by zmstone** [comment]
@dgriff1 yes.
